files/grest/rpc/01_cached_tables/stake_distribution_cache.sql

```diff
@@ -1,301 +1,238 @@
+
+  SELECT id INTO _last_active_stake_blockid FROM PUBLIC.BLOCK
+    WHERE epoch_no = _active_stake_epoch
+    ORDER BY block_no DESC LIMIT 1 ;
+
+  WITH 
+    accounts_with_delegated_pools AS (
+      SELECT DISTINCT ON (STAKE_ADDRESS.ID) stake_address.id as stake_address_id, stake_address.view as stake_address, pool_hash_id
+      FROM STAKE_ADDRESS
+        INNER JOIN DELEGATION ON DELEGATION.ADDR_ID = STAKE_ADDRESS.ID
+        WHERE
+          NOT EXISTS (
+            SELECT TRUE FROM DELEGATION D
+              WHERE D.ADDR_ID = DELEGATION.ADDR_ID AND D.ID > DELEGATION.ID
+          )
+          AND NOT EXISTS (
+            SELECT TRUE FROM STAKE_DEREGISTRATION
+              WHERE STAKE_DEREGISTRATION.ADDR_ID = DELEGATION.ADDR_ID
+                AND STAKE_DEREGISTRATION.TX_ID > DELEGATION.TX_ID
+          )
+    ),
+    pool_ids as (
+      SELECT awdp.stake_address_id,
+        pool_hash.view AS pool_id
+      FROM POOL_HASH
+        INNER JOIN accounts_with_delegated_pools awdp ON awdp.pool_hash_id = pool_hash.id
+    ),
+    account_active_stake AS (
+      SELECT awdp.stake_address_id, acsc.amount
+      FROM grest.account_active_stake_cache acsc
+        INNER JOIN accounts_with_delegated_pools awdp ON awdp.stake_address = acsc.stake_address
+        WHERE epoch_no = _active_stake_epoch
+    ),
+    account_delta_tx_ins AS (
+      SELECT awdp.stake_address_id, tx_in.tx_out_id AS txoid, tx_in.tx_out_index AS txoidx FROM tx_in
+        LEFT JOIN tx_out ON tx_in.tx_out_id = tx_out.tx_id AND tx_in.tx_out_index::smallint = tx_out.index::smallint
+        INNER JOIN accounts_with_delegated_pools awdp ON awdp.stake_address_id = tx_out.stake_address_id
+        WHERE tx_in.tx_in_id > (SELECT MAX(id) FROM tx WHERE block_id = _last_active_stake_blockid)
+    ),
+    account_delta_input AS (
+      SELECT tx_out.stake_address_id, COALESCE(SUM(tx_out.value), 0) AS amount
+      FROM account_delta_tx_ins
+        LEFT JOIN tx_out ON account_delta_tx_ins.txoid=tx_out.tx_id AND account_delta_tx_ins.txoidx = tx_out.index
+        INNER JOIN accounts_with_delegated_pools awdp ON awdp.stake_address_id = tx_out.stake_address_id
+        GROUP BY tx_out.stake_address_id
+    ),
+    account_delta_output AS (
+      SELECT awdp.stake_address_id, COALESCE(SUM(tx_out.value), 0) AS amount
+      FROM tx_out
+        INNER JOIN accounts_with_delegated_pools awdp ON awdp.stake_address_id = tx_out.stake_address_id
+        INNER JOIN tx ON tx.id = tx_out.tx_id
+        LEFT JOIN tx_in ON tx_out.tx_id = tx_in.tx_out_id AND tx_out.index::smallint = tx_in.tx_out_index::smallint
+      WHERE TX.BLOCK_ID > _last_active_stake_blockid
+      GROUP BY awdp.stake_address_id
+    ),
+    account_delta_rewards AS (
+      SELECT awdp.stake_address_id, COALESCE(SUM(reward.amount), 0) AS REWARDS
+      FROM REWARD
+        INNER JOIN accounts_with_delegated_pools awdp ON awdp.stake_address_id = reward.addr_id
+      WHERE REWARD.SPENDABLE_EPOCH >= _active_stake_epoch
```

Comment:
I think we are missing an upper boundary check here that `SPENDABLE_EPOCH` is not greater than `CURRENT_EPOCH`.

Reply:
That's interesting, I didn't think `spendable_epoch` for next epoch is recorded by dbsync yet :

``` sql
select from reward where spendable_epoch > (select max(no) from epoch) limit 10;
# --
# (0 rows)
```

Reply:
From mainnet DB just now:
```
cexplorer=# select max(no) from epoch;
 max
-----
 329
(1 row)

cexplorer=# select max(spendable_epoch) from reward;
 max
-----
 330
(1 row)
```

Reply:
Cool, wasnt aware of this one - it's a good outcome then